pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/BaseOffHeapMutableDictionary.java

```diff
@@ -0,0 +1,476 @@
+  private final int _sizeOfFirstBuf;
+
+  /**
+   * A class to hold all the objects needed for the reverse mapping.
+   */
+  private static class ValueToDictId {
+    // Each iBuf layout is as described in comments above.
+    private final List<IntBuffer> _iBufList;
+    // The map should be a concurrent one.
+    private final Map<Object, Integer> _overflowMap;
+
+    private ValueToDictId(List<IntBuffer> iBufList, Map<Object, Integer> overflowMap) {
+      _iBufList = iBufList;
+      _overflowMap = overflowMap;
+    }
+
+    private List<IntBuffer> getIBufList() {
+      return _iBufList;
+    }
+    private Map<Object, Integer> getOverflowMap() {
+      return _overflowMap;
+    }
+  }
+
+  private AtomicReference<ValueToDictId> _valueToDict = new AtomicReference<>();
+
+  protected BaseOffHeapMutableDictionary(int estimatedCardinality, int maxOverflowHashSize) {
+    final int initialRowCount = nearestPrime(estimatedCardinality);
+    _numEntries = 0;
+    _maxItemsInOverflowHash = maxOverflowHashSize;
+    ValueToDictId valueToDictId = new ValueToDictId(new ArrayList<IntBuffer>(0), new ConcurrentHashMap<Object, Integer>(0));
+    _valueToDict.set(valueToDictId);
+    _sizeOfFirstBuf = initialRowCount * NUM_COLUMNS;
+    if (!HEAP_FIRST || (maxOverflowHashSize == 0)) {
+      expand(_sizeOfFirstBuf);
+    }
+  }
+
+  @Override
+  public int indexOf(Object rawValue) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Object get(int dictionaryId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getLongValue(int dictionaryId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public double getDoubleValue(int dictionaryId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public float getFloatValue(int dictionaryId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int getIntValue(int dictionaryId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int length() {
+    return _numEntries;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return _numEntries == 0;
+  }
+
+  @Override
+  public void close() throws IOException {
+    doClose();
+
+    ValueToDictId valueToDictId = _valueToDict.get();
+    _valueToDict.set(null);
+    Map<Object, Integer> overflowMap = valueToDictId.getOverflowMap();
+    overflowMap.clear();
+    List<IntBuffer> iBufs = valueToDictId.getIBufList();
+    iBufs.clear();
+
+    for (PinotDataBuffer pinotDataBuffer : _pinotDataBuffers) {
+      pinotDataBuffer.close();
+    }
+    _pinotDataBuffers.clear();
+  }
+
+  private int nearestPrime(int size) {
+    for (int i = 0; i < PRIME_NUMBERS.length; i++) {
+      if (PRIME_NUMBERS[i] >= size) {
+        return PRIME_NUMBERS[i];
+      }
+    }
+    return PRIME_NUMBERS[PRIME_NUMBERS.length-1];
+  }
+
+  private IntBuffer expand(final int newSize) {
```

Comment:
that's why I suggested numEntries or targetNumEntries (you are also using _numEntries to track uniques)